core/src/main/scala/org/apache/spark/sql/delta/stats/StatisticsCollection.scala

```diff
@@ -389,7 +384,306 @@ trait StatisticsCollection extends DeltaLogging {
+        .map(_.nameParts)
+        .map { attributeNameParts =>
+          val commonPrefix = oldColumnPath.zip(attributeNameParts)
+            .takeWhile { case (left, right) => left == right }
+            .size
+          if (commonPrefix == oldColumnPath.size) {
+            newColumnPath ++ attributeNameParts.takeRight(attributeNameParts.size - commonPrefix)
+          } else {
+            attributeNameParts
+          }
+        }
+        .map(columnParts => UnresolvedAttribute(columnParts).name)
+      Map(
+        DeltaConfigs.DATA_SKIPPING_STATS_COLUMNS.key -> deltaStatsColumnsPath.mkString(",")
+      )
+    }.getOrElse(Map.empty[String, String])
+  }
+
+  /** Returns the configured set of columns to be used for stats collection on a table */
+  def configuredDeltaStatsColumnSpec(metadata: Metadata): DeltaStatsColumnSpec = {
+    val indexedColNamesOpt = DeltaConfigs.DATA_SKIPPING_STATS_COLUMNS.fromMetaData(metadata)
+    val numIndexedCols = DeltaConfigs.DATA_SKIPPING_NUM_INDEXED_COLS.fromMetaData(metadata)
+    indexedColNamesOpt.map { indexedColNames =>
+      DeltaStatsColumnSpec(parseDeltaStatsColumnNames(indexedColNames), None)
+    }.getOrElse {
+      DeltaStatsColumnSpec(None, Some(numIndexedCols))
+    }
+  }
+
+  /**
+   * Generates a filtered data schema for stats collection.
+   * Note: This method is matching the logical name of tables with the columns inside
+   * DeltaConfigs.DATA_SKIPPING_STATS_COLUMNS. The output of the filter schema is translated into
+   * physical name.
+   *
+   * @param schemaNames the full name path of all columns inside `schema`.
+   * @param schema the original data schema.
+   * @param statsColPaths the specific set of columns to collect stats on.
+   * @param mappingMode the column mapping mode of this statistics collection.
+   * @param parentPath the parent column path of `schema`.
+   * @return filtered schema
+   */
+  def filterSchema(
+      schemaNames: Seq[String],
+      schema: StructType,
+      statsColPaths: Seq[Seq[String]],
+      mappingMode: DeltaColumnMappingMode,
+      parentPath: Seq[String] = Seq.empty): Option[StructType] = {
+    // Find the unique column names at this nesting depth, each with its path remainders (if any)
+    val cols = statsColPaths.groupBy(_.head).mapValues(_.map(_.tail))
+    val newSchema = schema.flatMap { field =>
+      cols.get(field.name).flatMap { paths =>
+        field.dataType match {
+          case _ if paths.forall(_.isEmpty) =>
+            val fullPath = (parentPath:+ field.name).mkString(".")
+            def convertToPhysicalName(field: StructField): StructField = {
+              // If mapping mode is NoMapping or the dataSchemaName already contains the mapped
+              // column name, the schema mapping can be skipped.
+              if (mappingMode == NoMapping || schemaNames.contains(fullPath)) return field
+              val physicalName = field.metadata.getString("delta.columnMapping.physicalName")
```

Comment:
this constant should be defined somewhere. its likely this is being used multiple places.

Reply:
Done